src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/InlineCompletionHandler.cs

```diff
@@ -0,0 +1,176 @@
+            _requestInvoker = requestInvoker;
+            _projectionProvider = projectionProvider;
+            _documentMappingProvider = documentMappingProvider;
+            _formattingOptionsProvider = formattingOptionsProvider;
+
+            _logger = loggerProvider.CreateLogger(nameof(InlineCompletionHandler));
+        }
+
+        public async Task<VSInternalInlineCompletionList?> HandleRequestAsync(VSInternalInlineCompletionRequest request, ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            _logger.LogInformation($"Starting request for {request.TextDocument.Uri} at {request.Position}.");
+
+            if (!_documentManager.TryGetDocument(request.TextDocument.Uri, out var documentSnapshot))
+            {
+                return null;
+            }
+
+            var projectionResult = await _projectionProvider.GetProjectionAsync(documentSnapshot, request.Position, cancellationToken).ConfigureAwait(false);
+            if (projectionResult is null)
+            {
+                _logger.LogWarning($"Failed to find document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            if (projectionResult.LanguageKind != RazorLanguageKind.CSharp && projectionResult.LanguageKind != RazorLanguageKind.Html)
+            {
+                _logger.LogInformation($"Inline completions not supported for {projectionResult.LanguageKind}");
+                return null;
+            }
+
+            var inlineCompletionParams = new VSInternalInlineCompletionRequest
+            {
+                Context = request.Context,
+                TextDocument = new TextDocumentIdentifier { Uri = projectionResult.Uri },
+                Position = projectionResult.Position,
+            };
+
+            _logger.LogInformation($"Requesting inline completions for {projectionResult.Uri}.");
+
+            var serverKind = projectionResult.LanguageKind.ToLanguageServerKind();
+            var textBuffer = serverKind.GetTextBuffer(documentSnapshot);
+            var languageServerName = serverKind.ToLanguageServerName();
+            var response = await _requestInvoker.ReinvokeRequestOnServerAsync<VSInternalInlineCompletionRequest, VSInternalInlineCompletionList?>(
+                textBuffer,
+                VSInternalMethods.TextDocumentInlineCompletionName,
+                languageServerName,
+                inlineCompletionParams,
+                cancellationToken).ConfigureAwait(false);
+
+            if (!ReinvocationResponseHelper.TryExtractResultOrLog(response, _logger, languageServerName, out var result))
+            {
+                return null;
+            }
+
+            _logger.LogInformation("Received result, remapping.");
+
+            if (response?.Response == null)
+            {
+                return null;
```

Comment:
Is this worth logging about or does it just mean that LS didn't have anything for this spot?

Reply:
It means that it didn't have anything.  Will add a log